name: Continuous Integration

on:
  pull_request:
    branches:
      - main
  push:
    branches:
      - main

permissions:
  contents: read

jobs:
  test-typescript:
    name: TypeScript Tests
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        id: checkout
        uses: actions/checkout@v4

      - name: Setup Node.js
        id: setup-node
        uses: actions/setup-node@v4
        with:
          node-version-file: .node-version
          cache: npm

      - name: Install Dependencies
        id: npm-ci
        run: npm ci

      - name: Check Format
        id: npm-format-check
        run: npm run format:check

      - name: Lint
        id: npm-lint
        run: npm run lint

      - name: Test
        id: npm-ci-test
        run: npm run ci-test

  test-action-success:
    name: GitHub Action Test
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        id: checkout
        uses: actions/checkout@v4

      - name: Test Local Action
        id: test-action
        uses: ./
        with:
          up: true
          check: true

      - name: Print environment
        id: env
        run: printenv | sort

      - name: Print Output
        id: output
        run: echo "${{ toJSON(steps.test-action.outputs) }}"

  test-action-up-error:
    name: GitHub Action Test (failing 'omni up')
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          path: setup-action

      - name: Create empty git repository
        run: |
          git init
          git remote add origin https://example.com/test-repo.git

          cat >.omni.yaml <<EOF
          up:
            - this-is-not-a-supported-tool
          EOF

      - name: Test Local Action
        id: test-action
        continue-on-error: true
        uses: ./setup-action/
        with:
          up: true

      - name: Check if action errored
        env:
          ACTION_OUTCOME: ${{ steps.test-action.outcome }}
          ACTION_ERROR: ${{ steps.test-action.outputs.error }}
        run: |
          if [[ "$ACTION_OUTCOME" == "success" ]]; then
            echo "Action did not error"
            exit 1
          fi

          if [[ "$ACTION_ERROR" != "omni up "*" failed with exit code 1" ]]; then
            echo "Action did not error with the expected error message"
            exit 1
          fi

  test-action-check-error:
    name: GitHub Action Test (failing 'omni config check')
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          path: setup-action

      - name: Create empty git repository
        run: |
          git init
          git remote add origin https://example.com/test-repo.git

          cat >.omni.yaml <<EOF
          up: false  # Should be an object, the check should complain
          EOF

      - name: Test Local Action
        id: test-action
        continue-on-error: true
        uses: ./setup-action/
        with:
          check: true

      - name: Check if action errored
        env:
          ACTION_OUTCOME: ${{ steps.test-action.outcome }}
          ACTION_ERROR: ${{ steps.test-action.outputs.error }}
        run: |
          if [[ "$ACTION_OUTCOME" == "success" ]]; then
            echo "Action did not error"
            exit 1
          fi

          if [[ "$ACTION_ERROR" != "omni config check "*" failed with exit code 1" ]]; then
            echo "Action did not error with the expected error message"
            exit 1
          fi
